k2/csrc/weights.cc

```diff
@@ -68,11 +69,81 @@ WfsaWithFbWeights::WfsaWithFbWeights(const Fsa &fsa, const float *arc_weights,
+// 2002.
+void WfsaWithFbWeights::ComputeForwardWeights() {
+  auto num_states = fsa.NumStates();
+  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]);
+  std::fill_n(forward_state_weights.get(), num_states, kDoubleNegativeInfinity);
+  std::vector<double> rd(num_states, kDoubleNegativeInfinity);
+
+  auto add_weights_func = AddWeights();
+  const auto &arcs = fsa.arcs;
+  forward_state_weights[0] = 0;
+  rd[0] = 0;
+  for (std::size_t i = 0; i != arcs.size(); ++i) {
+    const auto &arc = arcs[i];
+    DCHECK_GE(arc.dest_state, arc.src_state);
+    auto src_weight = forward_state_weights[arc.src_state];
+    auto &dest_weight = forward_state_weights[arc.dest_state];
+
+    // will be `kNegativeInfinity` for non-accessible state
+    double r = rd[arc.src_state] + arc_weights[i];
+    // will be `dest_weight` for non-accessible state
+    double d = add_weights_func(dest_weight, r);
+    dest_weight = d;
+    // do not overwrite original `rd` value of `src_state`
+    // as we may use it in the next iteration
+    if (arc.dest_state != arc.src_state)
+      rd[arc.dest_state] = add_weights_func(rd[arc.dest_state], r);
+  }
+}
 
+void WfsaWithFbWeights::ComputeBackardWeights() {
   auto num_states = fsa.NumStates();
-  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]());
-  backward_state_weights = std::unique_ptr<double[]>(new double[num_states]());
-  // TODO(haowen): compute `forward/backward_state_weights`, also check
-  // `IsTopSorted(fsa)`
+  backward_state_weights = std::unique_ptr<double[]>(new double[num_states]);
+  std::fill_n(backward_state_weights.get(), num_states,
+              kDoubleNegativeInfinity);
+  std::vector<double> rd(num_states, kDoubleNegativeInfinity);
+
+  auto add_weights_func = AddWeights();
+  const auto &arcs = fsa.arcs;
+  int32_t final_state = fsa.FinalState();
+  backward_state_weights[final_state] = 0;
+  rd[final_state] = 0;
+  for (auto i = static_cast<int32_t>(arcs.size()) - 1; i >= 0; --i) {
+    const auto &arc = arcs[i];
+    DCHECK_GE(arc.dest_state, arc.src_state);
+    auto &src_weight = backward_state_weights[arc.src_state];
+    auto dest_weight = backward_state_weights[arc.dest_state];
+
+    double r = rd[arc.dest_state] + arc_weights[i];
+    double d = add_weights_func(src_weight, r);
+    src_weight = d;
+    // do not overwrite original `rd` value of `dest_state`
+    // as we may use it in the next iteration
+    if (arc.src_state != arc.dest_state)
+      rd[arc.src_state] = add_weights_func(rd[arc.src_state], r);
+  }
 }
+
+WfsaWithFbWeights::AddWeightsFun WfsaWithFbWeights::AddWeights() const {
+  switch (weight_type) {
+    case kMaxWeight:
+      return [](double w1, double w2) { return std::max(w1, w2); };
+      break;
+    case kLogSumWeight:
+      return [](double w1, double w2) { return LogAdd(w1, w2); };
+      break;
+    default:
+      LOG(FATAL) << "Unreachable code is executed!";
```

Comment:
You have to return something here to remove compiler warnings.

Reply:
OK